tests/test_discover_api.py

```diff
@@ -1,20 +1,23 @@
 import calendar
+import uuid
+from contextlib import ExitStack
 from datetime import datetime
 from functools import partial
+
 import simplejson as json
-import uuid
 
 from snuba import settings
 from snuba.datasets.factory import enforce_table_writer, get_dataset
-
-from tests.base import BaseApiTest, get_event
+from tests.base import BaseApiTest, dataset_manager, get_event
 
 
 class TestDiscoverApi(BaseApiTest):
     def setup_method(self, test_method):
-        # Setup both tables
-        super().setup_method(test_method, "events")
-        super().setup_method(test_method, "transactions")
+        super().setup_method(test_method)
+
+        self.__dataset_manager = ExitStack()
+        for dataset_name in ["events", "transactions"]:
```

Comment:
Did you by chance try to put one dataset only (the discover one) here? In theory that should return all the storages (transactions and events) and run the create statements for them.

Reply:
I did not earlier, but just did: 

```
Traceback (most recent call last):
  File "/Users/ted/workspaces/snuba/tests/test_discover_api.py", line 16, in setup_method
    super().setup_method(test_method, "discover")
  File "/Users/ted/workspaces/snuba/tests/base.py", line 168, in setup_method
    super().setup_method(test_method, dataset_name)
  File "/Users/ted/workspaces/snuba/tests/base.py", line 118, in setup_method
    self.table = enforce_table_writer(self.dataset).get_schema().get_table_name()
  File "/Users/ted/workspaces/snuba/snuba/datasets/factory.py", line 96, in enforce_table_writer
    ), f"Dataset{dataset} does not have a writable storage."
AssertionError: Dataset<snuba.datasets.discover.DiscoverDataset object at 0x10c4bbc50> does not have a writable storage.
```

I think your logic makes sense, but it seems like these tests will need some more cleanup here to support that. I didn't dig much beyond that error though (why the table must be writable, or why there must only be one table — I assume this is mostly legacy leftovers, and has something to do with test data insertion?)

In general I think that most of the base test classes should probably be removed and replaced with pytest fixtures when possible, there are probably a lot of subclasses that don't rely on the superclass behavior.

I added a comment in 9f1d2cb728d65c8d219edfb0df4f0c638ddadf94 so we don't accidentally emulate this behavior elsewhere.